### DISCUSSION TITLE
stuck when cross-compile fltk-sys for aarch64-unknown-linux-gnu from x64 linux

### DISCUSSION THREAD

**Discussion by cstkingkey** [fltk-rs/fltk-rs#1053] [General]
using cargo-cross https://fltk-rs.github.io/fltk-book/Cross-Compiling.html.
cargo cross build stucks when cross-compile fltk-sys for aarch64-unknown-linux-gnu from x64 linux.

it keeps showing  Building [=======================>   ] 69/76: fltk-sys(build).
No errors or fails come out.

And I can't find the build log.
 
native x64 building is ok.

**Comment by MoAlyousef** [comment]
You can try `cross build -vv --target=aarch64-unknown-linux-gnu` to see if there’s actual output. The build can take longer with cargo-cross since it uses docker and the build might not be parallelized. 

You might also want to ask in the cross discussions:
https://github.com/rust-embedded/cross/discussions

**Comment by cstkingkey** [comment]
it outputs as following without failing:

[fltk-sys 1.2.21] cargo:rerun-if-changed=build/android.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/bundled.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/link.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/main.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/source.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/utils.rs
[fltk-sys 1.2.21] cargo:rerun-if-env-changed=CC
[fltk-sys 1.2.21] cargo:rerun-if-env-changed=CXX
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/CMakeLists.txt
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_widget.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_group.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_input.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_window.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_button.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_box.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_menu.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_dialog.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_valuator.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_browser.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_misc.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_text.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_image.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_draw.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_table.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_surface.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_printer.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_utils.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_macros.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_lock.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_lock.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_new.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_widget.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_group.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_window.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_button.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_box.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_menu.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_dialog.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_valuator.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_browser.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_misc.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_text.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_image.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_input.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_draw.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_table.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_tree.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_surface.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_printer.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_font.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_utils.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_nswindow.m
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/FL_FLex/FL_Flex.H
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/FL_FLex/FL_Flex.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_vec.hpp
[fltk-sys 1.2.21] fatal: Not a git repository (or any of the parent directories): .git
[fltk-sys 1.2.21] CMAKE_TOOLCHAIN_FILE_aarch64-unknown-linux-gnu = None
[fltk-sys 1.2.21] CMAKE_TOOLCHAIN_FILE_aarch64_unknown_linux_gnu = None
[fltk-sys 1.2.21] TARGET_CMAKE_TOOLCHAIN_FILE = None
[fltk-sys 1.2.21] CMAKE_TOOLCHAIN_FILE = None
[fltk-sys 1.2.21] CMAKE_PREFIX_PATH_aarch64-unknown-linux-gnu = None
[fltk-sys 1.2.21] CMAKE_PREFIX_PATH_aarch64_unknown_linux_gnu = None
[fltk-sys 1.2.21] TARGET_CMAKE_PREFIX_PATH = None
[fltk-sys 1.2.21] CMAKE_PREFIX_PATH = None
[fltk-sys 1.2.21] CMAKE_aarch64-unknown-linux-gnu = None
[fltk-sys 1.2.21] CMAKE_aarch64_unknown_linux_gnu = None
[fltk-sys 1.2.21] TARGET_CMAKE = None
[fltk-sys 1.2.21] CMAKE = None
[fltk-sys 1.2.21] running: "cmake" "/root/.cargo/registry/src/github.com-1ecc6299db9ec823/fltk-sys-1.2.21/cfltk" "-G" "Ninja" "-DOPTION_USE_SYSTEM_LIBPNG=OFF" "-DOPTION_USE_SYSTEM_LIBJPEG=OFF" "-DOPTION_USE_SYSTEM_ZLIB=OFF" "-DCFLTK_LINK_IMAGES=ON" "-DOpenGL_GL_PREFERENCE=GLVND" "-DOPTION_USE_GL=OFF" "-DCFLTK_USE_OPENGL=OFF" "-DOPTION_USE_PANGO=ON" "-DCFLTK_SINGLE_THREADED=OFF" "-DCMAKE_EXPORT_COMPILE_COMMANDS=ON" "-DCFLTK_CARGO_BUILD=ON" "-DFLTK_BUILD_EXAMPLES=OFF" "-DFLTK_BUILD_TEST=OFF" "-DOPTION_LARGE_FILE=ON" "-DOPTION_USE_THREADS=ON" "-DOPTION_BUILD_HTML_DOCUMENTATION=OFF" "-DOPTION_BUILD_PDF_DOCUMENTATION=OFF" "-DCMAKE_INSTALL_PREFIX=/home/rust/src/target/aarch64-unknown-linux-gnu/debug/build/fltk-sys-cb86faa26047532b/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC" "-DCMAKE_C_COMPILER=/usr/bin/aarch64-linux-gnu-gcc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC" "-DCMAKE_CXX_COMPILER=/usr/bin/aarch64-linux-gnu-g++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC" "-DCMAKE_ASM_COMPILER=/usr/bin/aarch64-linux-gnu-gcc" "-DCMAKE_BUILD_TYPE=Release"

without ninja feature：
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/android.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/bundled.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/link.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/main.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/source.rs
[fltk-sys 1.2.21] cargo:rerun-if-changed=build/utils.rs
[fltk-sys 1.2.21] cargo:rerun-if-env-changed=CC
[fltk-sys 1.2.21] cargo:rerun-if-env-changed=CXX
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/CMakeLists.txt
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_widget.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_group.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_input.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_window.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_button.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_box.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_menu.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_dialog.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_valuator.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_browser.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_misc.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_text.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_image.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_draw.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_table.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_surface.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_printer.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_utils.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_macros.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/include/cfl_lock.h
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_lock.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_new.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_widget.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_group.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_window.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_button.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_box.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_menu.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_dialog.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_valuator.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_browser.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_misc.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_text.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_image.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_input.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_draw.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_table.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_tree.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_surface.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_printer.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_font.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_utils.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_nswindow.m
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/FL_FLex/FL_Flex.H
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/FL_FLex/FL_Flex.cpp
[fltk-sys 1.2.21] cargo:rerun-if-changed=cfltk/src/cfl_vec.hpp
[fltk-sys 1.2.21] fatal: Not a git repository (or any parent up to mount point /cargo)
[fltk-sys 1.2.21] Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).

**Comment by cstkingkey** [comment]
after rebuilt cross docker image, it somehow works.
But, it now give errors.
  -- Pango requires Xft but Xft library or headers could not be found.
  -- Please install Xft development files and try again or disable OPTION_USE_PANGO.
  -- Configuring incomplete, errors occurred!

with no-pango feature, it complains about misssing of windows.h
/cargo/registry/src/github.com-1ecc6299db9ec823/fltk-sys-1.2.21/cfltk/fltk/src/drivers/WinAPI/Fl_WinAPI_Screen_Driver.H:27:21: fatal error: windows.h: No such file or directory

So which package I need to install?

**Reply by MoAlyousef** [reply]
The same packages for linux need to be also on the container. There are instructions on the cross README, under custom image.

**Reply by cstkingkey** [reply]
I do have installed all packages listed in the setup section in the container. still not working

**Reply by cstkingkey** [reply]
turn out specific arch packages need to be installed and make pkg-config work
(by ln the arch specific pc files into /use/lib/pkgconfig, maybe there is nicer ways).

**Reply by MoAlyousef** [reply]
I have added some instructions back here:
https://fltk-rs.github.io/fltk-book/Cross-Compiling.html

**Reply by MoAlyousef** [reply]
You can pass PKG_CONFIG_LIBDIR=/usr/lib/aarch64-linux-gnu/pkgconfig prior to the cross command:
```
PKG_CONFIG_LIBDIR=/usr/lib/aarch64-linux-gnu/pkgconfig cross build --target=...
```
But yes it requires installing pkg-config:arm64

**Comment by cstkingkey** [comment]
when cross docker image have ninja installed, the building will stuck.

**Reply by MoAlyousef** [reply]
It seems it’s a recurrent issue with ninja by searching the issue tracker. Ninja would be used by default in fltk-rs if it’s found, that’s because the previous version of cmake-rs didn’t support parallel builds. I might change that in the coming version since cmake-rs 0.1.46 now supports parallel builds.

**Reply by MoAlyousef** [reply]
Version 1.2.22 no longer uses ninja by default (if it’s found)
